src/log-store/src/raft_engine/log_store.rs

```diff
@@ -0,0 +1,545 @@
+                            "Successfully purged logstore files, namespaces need compaction: {:?}",
+                            res
+                        );
+                    }
+                    Err(e) => {
+                        error!(e; "Failed to purge files in logstore");
+                    }
+                }
+            }
+        });
+        *self.cancel_token.lock().await = Some(token);
+        *self.gc_task_handle.lock().await = Some(handle);
+        self.started.store(true, Ordering::Relaxed);
+        info!("RaftEngineLogstore started with config: {:?}", self.config);
+        Ok(())
+    }
+
+    async fn stop(&self) -> Result<(), Self::Error> {
+        ensure!(self.started(), IllegalStateSnafu);
+        self.started.store(false, Ordering::Relaxed);
+        let handle = self
+            .gc_task_handle
+            .lock()
+            .await
+            .take()
+            .context(IllegalStateSnafu)?;
+        let token = self
+            .cancel_token
+            .lock()
+            .await
+            .take()
+            .context(IllegalStateSnafu)?;
+        token.cancel();
+        handle.await.context(WaitGcTaskStopSnafu)?;
+        info!("RaftEngineLogstore stopped");
+        Ok(())
+    }
+
+    /// Append an entry to logstore. Currently of existence of entry's namespace is not checked.
+    async fn append(&self, e: Self::Entry) -> Result<AppendResponse, Self::Error> {
+        ensure!(self.started(), IllegalStateSnafu);
+        let entry_id = e.id;
+        let mut batch = LogBatch::with_capacity(1);
+        batch
+            .add_entries::<MessageType>(e.namespace_id, &[e])
+            .context(AddEntryLogBatchSnafu)?;
+
+        self.engine
+            .write(&mut batch, self.config.sync_write)
+            .context(RaftEngineSnafu)?;
+        Ok(AppendResponse { entry_id })
+    }
+
+    /// Append a batch of entries to logstore. `RaftEngineLogstore` assures the atomicity of
+    /// batch append.
+    async fn append_batch(
+        &self,
+        ns: &Self::Namespace,
+        entries: Vec<Self::Entry>,
+    ) -> Result<Vec<Id>, Self::Error> {
+        ensure!(self.started(), IllegalStateSnafu);
+        let entry_ids = entries.iter().map(Entry::get_id).collect::<Vec<_>>();
+        let mut batch = LogBatch::with_capacity(entries.len());
+        batch
+            .add_entries::<MessageType>(ns.id, &entries)
+            .context(AddEntryLogBatchSnafu)?;
+        self.engine
+            .write(&mut batch, self.config.sync_write)
+            .context(RaftEngineSnafu)?;
+        Ok(entry_ids)
+    }
+
+    /// Create a stream of entries from logstore in the given namespace. The end of stream is
+    /// determined by the current "last index" of the namespace.
+    async fn read(
+        &self,
+        ns: &Self::Namespace,
+        id: Id,
+    ) -> Result<SendableEntryStream<'_, Self::Entry, Self::Error>, Self::Error> {
+        ensure!(self.started(), IllegalStateSnafu);
```

Comment:
What happens if the engine is stopped during the read process?

Reply:
The stop method just cancels background garbage collection task , so stopping the logstore while a stream is still being read is totally fine.